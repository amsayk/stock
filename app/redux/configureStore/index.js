// @flow

import { createStore, applyMiddleware, compose } from 'redux';

import { APP_NAME, DEBUG, INIT } from 'vars';

import pick from 'lodash.pick';

import array from 'redux/middleware/array';
import bufferActions from 'redux/middleware/buffer-actions';

import { createHashHistory } from 'history';
import { routerMiddleware, routerActions } from 'react-router-redux';
import locationHelperBuilder from 'redux-auth-wrapper/history4/locationHelper';

import { client as apolloClient } from 'apollo';

import { createSnackbarController } from 'components/Snackbar';

import thunk from 'redux-thunk';

import rootReducer from 'redux/reducers';

const history = createHashHistory();

const configureStore = (initialState?: any) => {
  // ======================================================
  // Middleware Configuration
  // ======================================================
  const middlewares = [
    bufferActions(),
    array,
    thunk.withExtraArgument({
      client: apolloClient,
      history,
      get snackbar() {
        return (
          store.snackbar || (store.snackbar = createSnackbarController(store))
        );
      },
    }),
    routerMiddleware(history),
  ];

  // Redux DevTools Configuration
  const actionCreators = {
    ...routerActions,
  };

  const composeEnhancers =
    (__DEV__ || DEBUG) && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
          name: APP_NAME,
          routerActions,
        })
      : compose;

  // ======================================================
  // Store Enhancers
  // ======================================================
  const enhancers = [applyMiddleware(...middlewares)];

  const enhancer = composeEnhancers(...enhancers);

  const store = createStore(rootReducer, enhancer);

  // Snackbar
  store.snackbar = createSnackbarController(store);

  // Initialize
  store.dispatch({ type: INIT });

  if (__DEV__) {
    if (module.hot) {
      module.hot.accept('../reducers', () => {
        store.replaceReducer(require('../reducers'));
      });
    }
  }

  return store;
};

export default {
  configureStore,
  history,
  locationHelper: locationHelperBuilder({}),
};
