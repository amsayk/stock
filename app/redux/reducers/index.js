import user from './user/reducer';
import app from './app/reducer';
import snackbar from './snackbar/reducer';
import notification from './notification/reducer';
import scrolling from './scrolling/reducer';
import toastr from './toastr/reducer';
import intl from './intl/reducer';

import { USER_LOGGED_OUT } from './user/constants';

import { combineReducers } from 'redux-immutable';

import { INIT } from 'vars';

import { routerReducer as router } from 'react-router-redux';

import { reducer as formReducer, actionTypes } from 'redux-form/immutable';

const reducers = {
  intl,
  app,
  notification,
  scrolling,
  user,
  snackbar,
  toastr,
  form: formReducer.plugin({
    login: (state, action) => {
      switch (action.type) {
        case USER_LOGGED_OUT:
          return state
            ? state.mergeDeep({
                values: {
                  password: undefined,
                },
              })
            : state;
        case actionTypes.SUBMIT_START: // Clear errors
          return state
            ? state.merge({
                error: null,
              })
            : state;
        default:
          return state;
      }
    },
    changePassword: (state, action) => {
      switch (action.type) {
        case actionTypes.SET_SUBMIT_SUCCEEDED: // Clear passwords on success
          return state
            ? state.merge({
                values: {},
              })
            : state;
        default:
          return state;
      }
    },
    changeEmail: (state, action) => {
      switch (action.type) {
        case actionTypes.SET_SUBMIT_SUCCEEDED: // Clear email on success
          return state
            ? state.merge({
                values: {},
              })
            : state;
        default:
          return state;
      }
    },
  }),
};

export default combineReducers({
  router,
  ...reducers,
});
