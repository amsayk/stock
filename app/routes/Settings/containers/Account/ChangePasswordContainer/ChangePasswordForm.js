import React from 'react';
import T from 'prop-types';
import Link from 'react-router-dom/Link';
import withRouter from 'react-router-dom/withRouter';

import { compose } from 'redux';

import { withApollo } from 'react-apollo';

import { locationHelper } from 'redux/configureStore';

import {
  reduxForm,
  Field,
  propTypes as reduxFormPropTypes,
  SubmissionError,
} from 'redux-form/immutable';

import isEmpty from 'isEmpty';

import { intlShape } from 'react-intl';

import messages from 'routes/Settings/messages';

import style from 'routes/Settings/styles';

import CurrentPasswordField from 'routes/Settings/components/CurrentPasswordField';
import NewPasswordField from 'routes/Settings/components/NewPasswordField';

import refreshCurrentUser from 'utils/refreshCurrentUser';

import { PATH_SETTINGS_BASE, PATH_LOGIN } from 'vars';

import MUTATION from './setPassword.mutation.graphql';

export class ChangePasswordForm extends React.Component {
  state = {
    changePasswordAtNextLogin: this.props.changePasswordAtNextLogin,
  };

  constructor(props, context) {
    super(props, context);

    this.onKeyDown = this._onKeyDown.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }
  _onKeyDown(e) {
    if (e.key === 'Enter' && e.shiftKey === false) {
      const submit = this.props.handleSubmit(this.onSubmit);
      submit();
    }
  }

  async onSubmit(data) {
    const {
      data: { setPassword: { user, errors } },
    } = await this.props.client.mutate({
      mutation: MUTATION,
      variables: {
        payload: {
          currentPassword: data.get('currentPassword'),
          newPassword: data.get('newPassword'),
        },
      },
    });

    if (!isEmpty(errors)) {
      throw new SubmissionError(errors);
    }

    refreshCurrentUser(user, this.props.dispatch);

    const { intl } = this.props;
    const { snackbar } = this.context;
    if (snackbar) {
      snackbar.show({
        duration: 2000,
        message: intl.formatMessage(messages.passwordChangeSuccessNotification),
      });
    }

    if (this.state.changePasswordAtNextLogin) {
      const redirectPath = locationHelper.getRedirectQueryParam(this.props);
      setTimeout(() => this.props.history.push(redirectPath || '/'), 100);
    }
  }

  render() {
    const { intl, handleSubmit, pristine, submitting } = this.props;
    return (
      <div className={style.content}>
        <h1 className={style.formHeading}>
          {intl.formatMessage(messages.titleChangePassword)}
        </h1>
        <div className={style.form}>
          <Field
            name="currentPassword"
            component={CurrentPasswordField}
            label={intl.formatMessage(messages.labelCurrentPassword)}
            onKeyDown={this.onKeyDown}
          />
          <div style={{ marginTop: 15 }} />
          <Field
            name="newPassword"
            component={NewPasswordField}
            label={intl.formatMessage(messages.labelNewPassword)}
            onKeyDown={this.onKeyDown}
          />
          <div className={style.formGroup}>
            <div className={style.offset2} />
            <div style={{ marginLeft: 15 }} className={style.buttonGroup}>
              <button
                onClick={handleSubmit(this.onSubmit)}
                disabled={pristine || submitting}
                className={style.changePasswordButton}
              >
                {intl.formatMessage(messages.changePassword)}
              </button>
              <Link to={PATH_SETTINGS_BASE} className={style.cancelButton}>
                {intl.formatMessage(messages.cancel)}
              </Link>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

ChangePasswordForm.contextTypes = {
  snackbar: T.shape({
    show: T.func.isRequired,
  }),
};

ChangePasswordForm.defaultProps = {};

ChangePasswordForm.propTypes = {
  ...reduxFormPropTypes,
  intl: intlShape.isRequired,
};

const Form = reduxForm({
  form: 'changePassword',
});

export default compose(withApollo, withRouter, Form)(ChangePasswordForm);
