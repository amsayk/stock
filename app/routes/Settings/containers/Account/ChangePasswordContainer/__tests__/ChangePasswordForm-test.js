import React from 'react';
import { compose } from 'redux';
import configureMockStore from 'redux-mock-store';
import { reduxForm } from 'redux-form/immutable';
import { IntlProvider, injectIntl } from 'react-intl';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import { fromJS } from 'immutable';

import { ChangePasswordForm } from '../ChangePasswordForm';

const client = {};

const Decorated = compose(injectIntl, reduxForm({ form: 'testForm' }))(
  ChangePasswordForm,
);

const middlewares = [];
const mockStore = configureMockStore(middlewares);

describe('change password form', () => {
  const store = mockStore(
    fromJS({
      form: {
        testForm: {},
      },
    }),
  );

  it('should render the snapshot', () => {
    const tree = renderer
      .create(
        <IntlProvider
          defaultLocale={'en'}
          locale={'en'}
          messages={{}}
          formats={{}}
        >
          <Provider store={store}>
            <Decorated client={client} />
          </Provider>
        </IntlProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
