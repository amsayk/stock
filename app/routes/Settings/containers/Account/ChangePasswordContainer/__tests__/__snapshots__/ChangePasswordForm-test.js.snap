exports[`change password form should render the snapshot 1`] = `
<div
  className="content">
  <h1
    className="formHeading">
    Change your password
  </h1>
  <div
    className="form">
    <div
      className="formGroup">
      <label
        className="label"
        htmlFor="currentPassword">
        Current password
      </label>
      <div
        className="inputWrapper">
        <input
          autoComplete="off"
          autoFocus={true}
          className="control"
          name="currentPassword"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          type="password"
          value="" />
        <div />
      </div>
    </div>
    <br />
    <div
      className="formGroup">
      <label
        className="label"
        htmlFor="newPassword">
        New password
      </label>
      <div
        className="inputWrapper">
        <input
          autoComplete="off"
          className="control"
          name="newPassword"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          type="password"
          value="" />
        <div />
      </div>
    </div>
    <div
      className="buttonGroup">
      <button
        className="changePasswordButton"
        disabled={true}
        onClick={[Function]}>
        Change password
      </button>
      <a
        className="cancelButton"
        onClick={[Function]}
        style={Object {}}>
        Cancel
      </a>
    </div>
  </div>
</div>
`;
